x-pack/plugins/security_solution/public/common/containers/use_search_strategy/index.tsx

```diff
@@ -0,0 +1,171 @@
+    timerange: transformTimerange,
+  } = getTransformChangesIfTheyExist({
+    factoryQueryType,
+    indices: defaultIndex,
+    filterQuery,
+    timerange,
+  });
+
+  return data.search.search<RequestBasicOptions, ResponseType>(
+    {
+      ...requestProps,
+      factoryQueryType: transformFactoryQueryType,
+      defaultIndex: transformIndices,
+      timerange: transformTimerange,
+      filterQuery,
+    },
+    {
+      strategy: 'securitySolutionSearchStrategy',
+      abortSignal: signal,
+    }
+  );
+};
+
+const searchComplete = <ResponseType extends IKibanaSearchResponse>(
+  props: UseSearchStrategyRequestArgs
+): Observable<ResponseType> => {
+  return search<ResponseType>(props).pipe(
+    filter((response) => {
+      return isErrorResponse(response) || isCompleteResponse(response);
+    })
+  );
+};
+
+export const useSearchStrategy = <QueryType extends FactoryQueryTypes>({
+  factoryQueryType,
+  initialResult,
+  errorMessage,
+}: {
+  factoryQueryType: QueryType;
+  /**
+   * `result` initial value. It is used until the search strategy returns some data.
+   */
+  initialResult: Omit<StrategyResponseType<QueryType>, 'rawResponse'>;
+  /**
+   * Message displayed to the user on a Toast when an erro happens.
+   */
+  errorMessage?: string;
+}) => {
+  const abortCtrl = useRef(new AbortController());
+  const { getTransformChangesIfTheyExist } = useTransforms();
+
+  const refetch = useRef<inputsModel.Refetch>(noop);
+  const { data } = useKibana().services;
+  const { addError } = useAppToasts();
+
+  const { start, error, result, loading } = useObservable<
+    [UseSearchStrategyRequestArgs],
+    StrategyResponseType<QueryType>
+  >(searchComplete);
+
+  useEffect(() => {
+    if (error != null) {
+      addError(error, {
+        title: errorMessage ?? i18n.DEFAULT_ERROR_SEARCH_STRATEGY(factoryQueryType),
+      });
+    }
+  }, [addError, error, errorMessage, factoryQueryType]);
+
+  const searchCb = useCallback(
+    (props: OptionalSignalArgs<StrategyRequestType<QueryType>>) => {
+      const asyncSearch = () => {
+        abortCtrl.current = new AbortController();
+        start({
+          ...props,
+          data,
+          factoryQueryType,
+          getTransformChangesIfTheyExist,
+          signal: abortCtrl.current.signal,
+        } as never); // This typescast is required because every StrategyRequestType instance has different fields.
+      };
+
+      abortCtrl.current.abort();
+      asyncSearch();
+
+      refetch.current = asyncSearch;
+    },
+    [data, start, factoryQueryType, getTransformChangesIfTheyExist]
+  );
+
+  useEffect(() => {
+    return () => {
+      abortCtrl.current.abort();
+    };
+  }, []);
+
+  const [formatedResult, inspect] = useMemo(
+    () => [
+      omit<StrategyResponseType<QueryType>, 'rawResponse'>('rawResponse', result),
+      getInspectResponse(result, {
+        dsl: [],
+        response: [],
+      }),
+    ],
+    [result]
+  );
```

Comment:
Maybe we can check if `result` is defined here, if not we should keep the same `initialResult` reference until the `result` is defined/complete, we have the `loading` at true meanwhile. The same for `inspect`, does it make sense to call `getInspectResponse` on an undefined `result`?

Reply:
> does it make sense to call getInspectResponse on an undefined result?

It doesn't make sense. I like your suggestion, let me implement it.